### PR TITLE
GH1743: Implements functionality for downloading NuGet packages

### LINKED ISSUE #1743: Implement functionality in Cake.NuGet for downloading packages
Implement functionality in Cake.NuGet for downloading packages instead of shelling out to `nuget.exe`

Benefits:
* Package installation works in `Cake.CoreCLR`. Today we do not allow since we can't be sure mono is actually installed on Linux.
* Allows us to work around nasty MAX_PATH issues with `nuget.exe` when path becomes too long
* More control over actual name of folder where package is extracted

### PR DESCRIPTION
**Pull request by mholo65**
- Fixes #1743

TODO:
- [x] Make this opt-in through configuration param.
- [ ] Add integration tests.
- [x] Opt-in Load Dependencies (both config and directive)

### CODE REVIEW

File: src/Cake.NuGet/Install/NuGetSourceRepositoryProvider.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace Cake.NuGet.Install
+{
+    internal sealed class NuGetSourceRepositoryProvider : ISourceRepositoryProvider
+    {
+        private static readonly string[] _defaultSources =
+        {
+            "https://api.nuget.org/v3/index.json"
```
Comment (gep13) [inline src/Cake.NuGet/Install/NuGetSourceRepositoryProvider.cs]: @mholo65 should this come from the `cake.config` file?

Comment (devlead) [reply src/Cake.NuGet/Install/NuGetSourceRepositoryProvider.cs]: Yes, I think it should.

Comment (mholo65) [reply src/Cake.NuGet/Install/NuGetSourceRepositoryProvider.cs]: Yes it should!

File: src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs [hunk 2]
```diff
@@ -65,6 +98,26 @@ public void Shouls_Register_The_NuGet_Package_Installer()
...
+                var fixture = new NuGetModuleFixture<NuGetPackageInstaller>();
+                fixture.Configuration.SetValue(Constants.NuGet.UseInProcessClient, bool.TrueString);
+                var module = fixture.CreateModule();
+
+                // When
+                module.Register(fixture.Registrar);
+
+                // Then
+                fixture.Registrar.Received(1).RegisterType<Install.NuGetPackageInstaller>();
+                fixture.Builder.Received(1).As<INuGetPackageInstaller>();
```
Comment (mholo65) [inline src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs]: @cake-build/cake-team can anyone tell me why this test is failing here?

**Comment by mholo65** [comment]
@devlead rebased.

File: src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs [hunk 3]
```diff
@@ -65,6 +98,26 @@ public void Shouls_Register_The_NuGet_Package_Installer()
...
                 fixture.Builder.Received(1).Singleton();
             }
+
+#pragma warning disable xUnit1004 // Test methods should not be skipped
+            [Fact(Skip = "Don't know why this is failing...")]
+#pragma warning restore xUnit1004 // Test methods should not be skipped
+            public void Should_Register_The_In_Process_NuGet_Package_Installer_If_Set_In_Configuration()
+            {
+                // Given
+                var fixture = new NuGetModuleFixture<NuGetPackageInstaller>();
```
Comment (devlead) [inline src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs]: This should be
```csharp
var fixture = new NuGetModuleFixture<Install.NuGetPackageInstaller>();
```

File: src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs (hunk 2 above)
Comment (devlead) [reply src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs]: see above comment

File: src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs (hunk 3 above)
Comment (mholo65) [reply src/Cake.NuGet.Tests/Unit/NuGetModuleTests.cs]: Ofc, stupid me looking at wrong place! Thank you!

**Comment by mholo65** [comment]
@devlead fixed

**Comment by devlead** [comment]
@mholo65 
Found one thing on Windows, the nuget client libs doesn't seem to be able to use sources with encrypted passwords. 

And it doesn't respect if you place a NuGet.config in repo root/tools folder/or cake folder. 

Interestingly the source it complains about is disabled, so it shouldn't be used, if i remove the disabled sources and only have 
```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="NuGetV3" value="https://api.nuget.org/v3/index.json" />
  </packageSources>
</configuration>
```

Feels like an issue with the NuGet client libs.

I don't get the same errors as I got on Linux so I'll rerun the tests there if it was something temporary.

**Comment by mholo65** [comment]
@devlead exception should be fixed... Will re-investigate re-download. If nothing helps, best would be to set `PackageSaveMode` to `NuPkg`. Any objections, @gep13?

**Comment by devlead** [comment]
@mholo65 I can confirm exception is sorted and it now works for me the first time.
But the second time I execute Cake with tools & addins restored in place it fails with:
```Error: System.AggregateException: One or more errors occurred. ---> System.InvalidOperationException: Package 'gitreleasemanager 0.5.0' is not found in the following primary source(s): '/pullrequest/tools'. Please verify all your online package sources are available (OR) package id, version are specified correctly.```

**Comment by mholo65** [comment]
@devlead I blame NuGet client libs and awful xplat support when it comes to file system. This breaks support for nuspec package save mode since nuspec is stored to disk as normalized (e.g. filename to lower invariant). But then when it checks if package is installed, it expects to find nuspec in non-normalized form.

**Comment by mholo65** [comment]
@devlead @gep13 rebased. Please take it for a spin.

**Comment by devlead** [comment]
@mholo65  your changes have been merged, thanks for your contribution 👍

**Pull request merged**